log/loggers.go

```diff
@@ -5,248 +5,204 @@ import (
 	mu.RLock()
 	defer mu.RUnlock()
 	fields := sl.getFields()
-	if fields == nil {
-		return
-	}
-	if fields.warn {
-		fields.output.StageLogEvent(func() string { return data },
-			fields.logger.WarnHeader,
-			fields.name,
-			fields.logger.Spacer,
-			fields.logger.TimestampFormat,
-			fields.logger.ShowLogSystemName,
-			fields.logger.BypassJobChannelFilledWarning)
-	}
-	logFieldsPool.Put(fields)
+	fields.stageln(fields.logger.ErrorHeader, v...)
 }
 
-// Warnln takes a pointer subLogger struct & interface formats and sends to StageLogEvent
-func Warnln(sl *SubLogger, v ...interface{}) {
+// Errorf takes a pointer subLogger struct, string and interface formats sends to StageLogEvent
+func Errorf(sl *SubLogger, data string, v ...interface{}) {
 	mu.RLock()
 	defer mu.RUnlock()
 	fields := sl.getFields()
-	if fields == nil {
-		return
+	sl.getFields().stagef(fields.logger.ErrorHeader, data, v...)
+}
+
+func displayError(err error) {
+	if err != nil {
+		log.Printf("Logger write error: %v\n", err)
 	}
-	if fields.warn {
-		fields.output.StageLogEvent(func() string { return fmt.Sprintln(v...) },
-			fields.logger.WarnHeader,
-			fields.name,
-			fields.logger.Spacer,
-			fields.logger.TimestampFormat,
-			fields.logger.ShowLogSystemName,
-			fields.logger.BypassJobChannelFilledWarning)
+}
+
+// enabled checks if the log level is enabled
+func (l *Fields) enabled(header string) string {
+	switch header {
+	case l.logger.InfoHeader:
+		if l.info {
+			return "info"
+		}
+	case l.logger.WarnHeader:
+		if l.warn {
+			return "warn"
+		}
+	case l.logger.ErrorHeader:
+		if l.error {
+			return "error"
+		}
+	case l.logger.DebugHeader:
+		if l.debug {
+			return "debug"
+		}
+	}
+	return ""
+}
+
+// stage stages a log event
+func (l *Fields) stage(header string, deferFunc deferral) {
+	if l == nil {
+		return
 	}
-	logFieldsPool.Put(fields)
+	if level := l.enabled(header); level != "" {
+		l.output.StageLogEvent(deferFunc,
+			header,
+			l.name,
+			l.logger.Spacer,
+			l.logger.TimestampFormat,
+			l.botName,
+			level,
+			l.logger.ShowLogSystemName,
+			l.logger.BypassJobChannelFilledWarning,
+			l.structuredLogging,
+			l.structuredFields)
+	}
+	logFieldsPool.Put(l)
 }
 
-// Warnf takes a pointer subLogger struct, string and interface formats sends to StageLogEvent
-func Warnf(sl *SubLogger, data string, v ...interface{}) {
+// stageln stages a log event
+func (l *Fields) stageln(header string, a ...interface{}) {
+	l.stage(header, func() string { return fmt.Sprint(a...) })
+}
+
+// stagef stages a log event
+func (l *Fields) stagef(header, format string, a ...interface{}) {
+	l.stage(header, func() string { return fmt.Sprintf(format, a...) })
+}
+
+// WithFields allows the user to add custom fields to a structured log output
+// NOTE: If structured logging is disabled, this function will do not add
+// new fields to the log output.
+func WithFields(sl *SubLogger, structuredFields ExtraFields) *Fields {
```

Comment:
This ~fields~ feels like trying to avoid making `DebugWithFields`, `InfoWithFields`, `ErrorWithFields`, `WarnWithFields`. Wouldn't the creation of those fix the issue you mentioned regarding holding onto a sublogger's `Fields` while it could potentially be inside the jobs pool?

Reply:
Yeah this works I will go with that 🚀 much safer.

Reply:
code is safu